xds/csds/csds.go

```diff
@@ -0,0 +1,317 @@
+	_ "google.golang.org/grpc/xds/internal/client/v3" // Register v3 xds_client.
+)
+
+// xdsClientInterface contains methods from xdsClient.Client which are used by
+// the server. This is useful for overriding in unit tests.
+type xdsClientInterface interface {
+	DumpLDS() (string, map[string]client.UpdateWithMD)
+	DumpRDS() (string, map[string]client.UpdateWithMD)
+	DumpCDS() (string, map[string]client.UpdateWithMD)
+	DumpEDS() (string, map[string]client.UpdateWithMD)
+	BootstrapConfig() *bootstrap.Config
+	Close()
+}
+
+var (
+	logger       = grpclog.Component("xds")
+	newXDSClient = func() (xdsClientInterface, error) {
+		return client.New()
+	}
+)
+
+// ClientStatusDiscoveryServer implementations interface ClientStatusDiscoveryServiceServer.
+type ClientStatusDiscoveryServer struct {
+	// xdsClient will always be the same in practise. But we keep a copy in each
+	// server instance for testing.
+	xdsClient xdsClientInterface
+}
+
+// NewClientStatusDiscoveryServer returns an implementation of the CSDS server that can be
+// registered on a gRPC server.
+func NewClientStatusDiscoveryServer() (*ClientStatusDiscoveryServer, error) {
+	ret := &ClientStatusDiscoveryServer{}
+	xdsC, err := newXDSClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create xds client: %v", err)
+	}
+	ret.xdsClient = xdsC
+	return ret, nil
+}
+
+// StreamClientStatus implementations interface ClientStatusDiscoveryServiceServer.
+func (s *ClientStatusDiscoveryServer) StreamClientStatus(stream v3statuspb.ClientStatusDiscoveryService_StreamClientStatusServer) error {
+	for {
+		req, err := stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		resp, err := s.buildClientStatusRespForReq(req)
+		if err != nil {
+			return err
+		}
+		if err := stream.Send(resp); err != nil {
+			return err
+		}
+	}
+}
+
+// FetchClientStatus implementations interface ClientStatusDiscoveryServiceServer.
+func (s *ClientStatusDiscoveryServer) FetchClientStatus(_ context.Context, req *v3statuspb.ClientStatusRequest) (*v3statuspb.ClientStatusResponse, error) {
+	return s.buildClientStatusRespForReq(req)
+}
+
+// buildClientStatusRespForReq fetches the status from the client, and returns
+// the response to be sent back to client.
+//
+// If it returns an error, the error is a status error.
+func (s *ClientStatusDiscoveryServer) buildClientStatusRespForReq(req *v3statuspb.ClientStatusRequest) (*v3statuspb.ClientStatusResponse, error) {
+	// Field NodeMatchers is unsupported, by design
+	// https://github.com/grpc/proposal/blob/master/A40-csds-support.md#detail-node-matching.
+	if len(req.NodeMatchers) != 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "node_matchers are not supported, request contains node_matchers: %v", req.NodeMatchers)
+	}
+	if s.xdsClient == nil {
+		// This should never happen.
+		return nil, status.Errorf(codes.FailedPrecondition, "xds client is nil")
+	}
```

Comment:
We don't need this check anymore, right?

Reply:
Removed. Should be safe now.